std/algorithm/iteration.d

```diff
@@ -4289,3 +4304,633 @@ unittest
+
+Returns:
+    Forward range which yields the product items
+*/
+auto cartesianPower(Range)(Range r, size_t repeat = 1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in
+{
+    assert(repeat >= 1, "Invalid repeat");
+}
+body
+{
+    import std.array: array;
+    import std.range: Indexed;
+    static struct CartesianPower
+    {
+    private:
+        Range _r;
+        size_t[] _state, _stateBack;
+
+        size_t _repeat, _nrElements;
+        size_t _max_states;
+        size_t _pos, _posBack;
+
+    public:
+        this(Range r, size_t repeat)
+        {
+            import std.math: pow;
+
+            _r = r;
+            // cache number of elements
+            _nrElements = r.length;
+            _repeat = repeat;
+
+            // set initial state and calculate max possibilities
+            if (_nrElements > 0)
+            {
+                _max_states = pow(_nrElements, _repeat);
+                if (_max_states > 0)
+                {
+                    _state = new size_t[](repeat);
+                    _stateBack = new size_t[](repeat);
+                    _stateBack[] = _nrElements -1;
+                }
+            }
+        }
+
+        @property typeof(this) save()
+        {
+            // we need to save more information about the current state
+            typeof(this) c = this;
+            c._state = _state.dup;
+            return c;
+        }
+
+        @property Indexed!(Range, size_t[]) front()
+        {
+            import std.range : indexed;
+            return _r.indexed(_state);
+        }
+
+        void popFront()
+        {
+            // check whether we have reached the end
+            if (empty) return;
+            _pos++;
+
+            /*
+            * Bitwise increment - starting from back
+            * It works like adding 1 in primary school arithmetic.
+            * If a block has reached the number of elements, we reset it to
+            * 0, and continue to increment, e.g. for n = 2:
+            *
+            * [0, 0, 0] -> [0, 0, 1]
+            * [0, 1, 1] -> [1, 0, 0]
+            */
+            foreach_reverse (i, ref el; _state)
+            {
+                ++el;
+                if (el < _nrElements)
+                    break;
+
+                el = 0;
+            }
+        }
+        @property Indexed!(Range, size_t[]) back()
+        {
+            import std.range : indexed;
+            return _r.indexed(_stateBack);
+        }
+
+        void popBack()
+        {
+            // check whether we have reached the end
+            if (empty) return;
+            _posBack++;
+            foreach_reverse (i, ref el; _stateBack)
+            {
+                if (el == 0)
+                    el = _nrElements - 1;
+                else
+                {
+                    --el;
+                    break;
+                }
+            }
+        }
+
+        @property size_t length()
+        {
+            return _max_states;
```

Comment:
The length must shrink as the range is consumed from the front or back; it doesn't seem to do that and it's not tested